Add environment variable expansion option for mirrors.yaml

kwryankrattiger opened the pull request:
Currently Spack expects either global mirror environment (S3) configuration to authenticate mirror access or injecting secrets into configs making them unshareable. To improve this a new user credentials file is introduced as well as environment variable expansion. This should make it easier for users to share configurations without risking about mixing in secrets.

kwryankrattiger committed:
Add environment variable expansion options

Currently Spack expects either global mirror environment (S3)
configuration to authenticate mirror access or injecting secrets into
configs making them unshareable. To improve this environment variable
options for passing credentials were added. This should make it easier
for users to share configurations without risking about mixing in secrets.

wdconinc commented:
It would be useful to update the documentation so it reflects this capability.

kwryankrattiger commented:
> It would be useful to update the documentation so it reflects this capability.

Yes, I was going to get review on the feature interface first and then update the docs according to the consensus.

tgamblin reviewed (commented):
My main thought for this is: AWS already has mechanisms for storing this type of stuff in well known locations.  I think you know this b/c you're calling this `~/.spack/credentials`, which is very similar to `~/.aws/credentials`

e.g., you can name your IAM profiles and put them in `~/.aws/credentials` like this:

```ini
[tgamblin]
aws_access_key_id: XXXXXXXXXXXXXXXXXXXX
aws_secret_access_key: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

[kitware]
aws_access_key_id: XXXXXXXXXXXXXXXXXXXX
aws_secret_access_key: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

[e4s-cache]
aws_access_key_id: XXXXXXXXXXXXXXXXXXXX
aws_secret_access_key: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

[spack-bootstrap-mirror-upload]
aws_access_key_id: XXXXXXXXXXXXXXXXXXXX
aws_secret_access_key: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
```

The tooling for AWS (boto3, awscli, etc.) knows how to use profiles (`tgamblin`, `e4s-cache`, etc.) from that file.  

So my suggestion is that we should just leverage that.  Users can put their crews in `~/.aws/credentials` (where they likely already have them for other reasons), and allow Spack to accept a profile name where it would currently accept a credential (basically just add support for e.g., `profile: e4s-cache` in our configs).  Then ensure that the profile we want to use is passed correctly to `boto3` or `awscli` underneath.

This way, I think we don't even have to know how to parse the AWS credential file. We can just tell *their* tools what part of it to look at.

If we do something similar for GCP, etc. we can leverage their credential systems, as well, and not wrap 18 different types of credentials.

Thoughts?

kwryankrattiger commented:
AWS and GCS provide, and certainly inspired, the idea of having a `.spack/credentials` configuration file.

Spack mirror clients such as OCI do not provide this type of global configuration file, and the current mechanism requires putting privileged credentials in the `mirrors.yaml` to authenticate the OCI cache. This is not ideal.

Having a common mechanisms for passing mirror credentials to Spack makes it possible to configure all Spack mirrors in a common way rather than relying solely on mirror type specific mechanisms which may or may not exist. This does not override or replace mirror type specific mechanisms, and it shouldn't as configuring credentials twice is annoying.

haampie commented:
For OCI many tools use `~/.docker/config.json`, but it is not common on macOS, and I think it only supports one user/pass pair per domain

kwryankrattiger commented:
I like the idea of requiring additional keywords to denote `variable`/`string`. Also being able to point at a file directly in the config seems reasonable to me.

On storing credentials to disk...yeah, not the most ideal solution. 100% that feature should _not_ be used in a CI context, which is true for any credential store for any application (AWS/docker/etc.). The only reason it is proposed is it is commonly done and is just a notch better than storing it in files that are more likely to be shared with `chmod 777`.

I am happy to completely drop it for the option to point at a credentials file in the config, which is at very least not easily searchable by "don't really trust" processes that may be looking at credentials files.

Simple format for a Spack credentials file?
```
<profile|pair-key|token>[:<pair-value>]
```

kwryankrattiger commented:
Okay. I have had some offline conversation on this. The new plan is to deprecate credentials as plain text strings, no file option, and `variable: MY_VAR` expansion as the preferred way.

This eliminates the can of worms that is "Spack is storing my secrets on disk" and allows per-mirror type configuration to continue to use the mirror provider authentication modes however they are.

kwryankrattiger commented on lib/spack/spack/schema/mirrors.py:
```diff
@@ -10,19 +10,60 @@
+                "type": "array",
+                "items": {"minItems": 2, "maxItems": 2, **string_or_variable},
+            },  # deprecated
+            {
```
This isn't strictly required, but I think it makes sense as a pair is two things and being able to clearly list them in the config is useful.

I do not have an option for just listing the `secret` in the new object format.

kwryankrattiger commented on lib/spack/spack/schema/mirrors.py:
```diff
@@ -10,19 +10,60 @@
     "profile": {"type": ["string", "null"]},
     "endpoint_url": {"type": ["string", "null"]},
+    "access_token": {"type": ["string", "null"]},  # deprecated
+    "access_token_variable": {"type": ["string", "null"]},
```
Replace the `access_token` with `access_token_variable`. The alternate syntax here was too many newlines for a field that wasn't going to get any additional options. If we someday want to add `file:` here maybe worth revisiting, but I think the current consensus is we don't want to support secrets on the FS.

kwryankrattiger commented on lib/spack/spack/schema/mirrors.py:
```diff
@@ -70,3 +113,37 @@
+
+    errors = []
+
+    def check_access_pair(name, section):
```
I am open to suggestions for a better way to do this.

kwryankrattiger commented:
@scottwittenburg can you check this out too

kwryankrattiger committed:
Add CLI support for variable credentials

kwryankrattiger commented:
More grouping for help output. Make arguments of same purpose mutually exclusive of `<arg>` vs `<arg>-variable`.

 ```
 usage: spack mirror add [-h] [--scope {defaults,system,site,user,command_line}[/PLATFORM] or env:ENVIRONMENT] [--type {binary,source}] [--autopush] [--unsigned | --signed]
                        [--s3-access-key-id S3_ACCESS_KEY_ID | --s3-access-key-id-variable S3_ACCESS_KEY_ID_VARIABLE]
                        [--s3-access-key-secret S3_ACCESS_KEY_SECRET | --s3-access-key-secret-variable S3_ACCESS_KEY_SECRET_VARIABLE]
                        [--s3-access-token S3_ACCESS_TOKEN | --s3-access-token-variable S3_ACCESS_TOKEN_VARIABLE] [--s3-profile S3_PROFILE] [--s3-endpoint-url S3_ENDPOINT_URL]
                        [--oci-username OCI_USERNAME | --oci-username-variable OCI_USERNAME_VARIABLE] [--oci-password OCI_PASSWORD | --oci-password-variable OCI_PASSWORD_VARIABLE]
                        mirror url

positional arguments:
  mirror                mnemonic name for mirror
  url                   url of mirror directory from 'spack mirror create'

optional arguments:
  --autopush            set mirror to push automatically after installation
  --scope {defaults,system,site,user,command_line}[/PLATFORM] or env:ENVIRONMENT
                        configuration scope to modify
  --signed              require signing and signature verification when pushing and installing from this build cache
  --type {binary,source}
                        specify the mirror type: for both binary and source use `--type binary --type source` (default)
  --unsigned            do not require signing and signature verification when pushing and installing from this build cache
  -h, --help            show this help message and exit

S3 Connection:
  --s3-access-key-id S3_ACCESS_KEY_ID
                        ID string to use to connect to this S3 mirror
  --s3-access-key-id-variable S3_ACCESS_KEY_ID_VARIABLE
                        Environment variable containing ID string to use to connect to this S3 mirror
  --s3-access-key-secret S3_ACCESS_KEY_SECRET
                        secret string to use to connect to this S3 mirror
  --s3-access-key-secret-variable S3_ACCESS_KEY_SECRET_VARIABLE
                        Environment variable containing secret string to use to connect to this S3 mirror
  --s3-access-token S3_ACCESS_TOKEN
                        access token to use to connect to this S3 mirror
  --s3-access-token-variable S3_ACCESS_TOKEN_VARIABLE
                        Environment variable containing access token to use to connect to this S3 mirror
  --s3-endpoint-url S3_ENDPOINT_URL
                        endpoint URL to use to connect to this S3 mirror
  --s3-profile S3_PROFILE
                        S3 profile name to use to connect to this S3 mirror

OCI Connection:
  --oci-password OCI_PASSWORD
                        password to use to connect to this OCI mirror
  --oci-password-variable OCI_PASSWORD_VARIABLE
                        Environment variable containing password to use to connect to this OCI mirror
  --oci-username OCI_USERNAME
                        username to use to connect to this OCI mirror
  --oci-username-variable OCI_USERNAME_VARIABLE
                        Environment variable containing username to use to connect to this OCI mirror
```

haampie commented:
This is broken:

```
$ spack 
$ spack mirror add --oci-username x --oci-password-variable y ghcr oci://ghcr.io/x/y
$ spack mirror set --oci-password-variable z ghcr
==> No changes made to mirror ghcr.
```

haampie commented on lib/spack/spack/cmd/common/arguments.py:
```diff
@@ -581,23 +581,51 @@ def add_concretizer_args(subparser):
+        group.add_argument(arg, **kwargs)
+        # Update help string
+        if "help" in kwargs:
+            kwargs["help"] = "Environment variable containing " + kwargs["help"]
```
```suggestion
            kwargs["help"] = f"environment variable containing {kwargs['help']}"
```

minor but `e` for consistency

haampie commented on lib/spack/spack/mirror.py:
```diff
@@ -271,11 +284,57 @@ def get_url(self, direction: str) -> str:
+        elif isinstance(pair, dict):
+            id_ = os.environ.get(pair["id_variable"]) if "id_variable" in pair else pair["id"]
+            secret = os.environ.get(pair["secret_variable"])
+            return (str(id_), str(secret)) if id_ and secret else None
```
```suggestion
            return (id_, secret) if id_ and secret else None
```

kwryankrattiger replied:
I originally had that, but I think mypy was complaining about types.

kwryankrattiger replied:
hmm, I ran the style checks again and it didn't complain this time, removed since it isn't very nice to look at.

haampie commented on lib/spack/spack/schema/mirrors.py:
```diff
@@ -10,19 +10,59 @@
+        "oneOf": [
+            {
+                "type": "array",
+                "items": {"minItems": 2, "maxItems": 2, **string_or_variable},
```
Why `string_or_variable`, you've just introduced a dict option below. Pick one :p

Also `null` was there for backward compatibility, don't delete it.

kwryankrattiger replied:
The only reason is to allow a variable name and a static password (see `cmd/mirror.py`). That case doesn't really make sense so I think dropping it is probably fine, I was getting a bit overly pedantic.

I will make that case an error and drop the string-or-variable.

haampie reviewed (changes requested):
I mostly reviewed it, but I think enough things to fix for now.

kwryankrattiger commented:
> This is broken, which means tests are missing: ...

So I looked and it is not actually broken, you must always specify both the username name and the password when updating using `set` or `set-url`. We can change that in the `Mirror._update_connection_dict` but the logic will temporarily need to be a bit more ugly.

ref.
```
if args.oci_username and args.oci_password:
```

kwryankrattiger committed:
Allow set/set-url to overwrite a single credential in an access pair

Add a lot more validation in the testing of setting access pairs on the
command line. Exhaustively test every combination and ensure expected
output.

haampie commented:
Still some usability issues. It by default creates deprecated output:

```
$ spack env create .
$ spack -e . mirror add foo oci://example.com/bar/baz
$ spack -e . mirror set foo --oci-username x --oci-password y

$ spack -e . mirror list
==> Warning: foo: Using access_pair with a list is deprecated, prefer id/secret_variable keys
==> Warning: foo: Secret part of access pair should be a variable
...
```

then following the second bit of the advise:

```
$ spack -e . mirror set foo --oci-username x --oci-password-variable y
==> Warning: foo: Using access_pair with a list is deprecated, prefer id/secret_variable keys
==> Warning: foo: Secret part of access pair should be a variable
==> No changes made to mirror foo.
```

kwryankrattiger committed:
Fix converting between mirror credential list <-> dict

kwryankrattiger commented:
> Still some usability issues. It by default creates deprecated output:

So, if your secret part is not a variable, it must use the deprecated output. The `--oci-password` and `--s3-access-key-secret` are also deprecated CLI options. The `deprecated` feature in `add_argument` was not working correctly so the warning doesn't show up there.

```
==> Error: __init__() got an unexpected keyword argument 'deprecated'
```

> then following the second bit of the advise doesn't work:

Fixed. This was introduced when adding the ability to set only one credential parameter in `set` and `set-url`.

> actually, don't put warnings in def update, we've never done that. A generic warning triggers when saving changes to mirrors.yaml. The current way is too noisy.

I agree, see my comment that I am looking for a better way to do this. Thinking about this more I think I can move this to the `_update_connection_dict`

kwryankrattiger committed:
Move access_pair deprecation warning to the mirror update

kwryankrattiger commented:
@haampie I moved the warning to the mirror update. Testing it locally it seems to work nicely now. I am working on adding the tests to handle the deprecated conversion case and printing deprecation warnings.

kwryankrattiger committed:
More tests and warnings

kwryankrattiger committed:
Update deprecated warning output to be more helpful

kwryankrattiger committed:
Add validation to early exit if mirror is not properly configured

haampie commented:
tests are failing

kwryankrattiger committed:
Drop requirement for OCI credentials

kwryankrattiger commented:
> tests are failing

@haampie 
Fixed.

haampie commented on lib/spack/spack/mirror.py:
```diff
@@ -154,8 +154,68 @@ def push_url(self):
         """Get the valid, canonicalized fetch URL"""
         return self.get_url("push")
 
+    def validate(self, direction="push"):
```
Probably better is `ensure_credentials_set` or something. Any function that returns None and raises is called ensure_* elsewhere.

Also add type hints.

And drop the comment about "URL is accessible". The only way of knowing for a fact is successfully pushing...

kwryankrattiger replied:
Sure, I originally had it return bool but decided to raise instead and reuse the `MirrorError`.l

For `s3`, and I think also`oci`, mirrors you can make a request to stat if the bucket/prefix exists.

kwryankrattiger replied:
We could also do a test write here on push and a test read on fetch..I'll drop the comment but put it on my backlog

haampie commented on lib/spack/spack/mirror.py:
```diff
@@ -154,8 +154,68 @@ def push_url(self):
+
+        # Verify that the credentials that are variables expand
+        if access_pair and isinstance(access_pair, dict):
+            if "id_variable" in access_pair:
+                if access_pair["id_variable"] not in os.environ:
```
have you heard of `and` ;) same another 2 times.

kwryankrattiger replied:
The original line was too long :p I can make it an `and` now

haampie commented on lib/spack/spack/mirror.py:
```diff
@@ -154,8 +154,68 @@ def push_url(self):
+        # TODO: Verify the URL is accessible
+
+        if errors:
+            msg = f"Could not validate push configuruation for mirror {self.name}:\n\t"
```
hard-coded `push`, but it is a function argument.

haampie commented on lib/spack/spack/mirror.py:
```diff
@@ -154,8 +154,68 @@ def push_url(self):
+
+        if errors:
+            msg = f"Could not validate push configuruation for mirror {self.name}:\n\t"
+            msg += "\n\t".join(errors)
```
I think in most places `    ` is used instead of `\t`.

haampie commented on lib/spack/spack/mirror.py:
```diff
@@ -271,11 +331,57 @@ def get_url(self, direction: str) -> str:
+        return creddict
+
+    @staticmethod
+    def _extract_credential_value(tok):
```
Can you add typehints

kwryankrattiger replied:
oh, this whole thing is not longer needed since we dropped the `- variable: FOO` option

haampie commented on lib/spack/spack/mirror.py:
```diff
@@ -271,11 +331,57 @@ def get_url(self, direction: str) -> str:
+
+    def get_access_pair(self, direction: str) -> Optional[Tuple[str, str]]:
+        pair = self._get_value("access_pair", direction)
+        if type(pair) in (tuple, list):
```
any reason to use type(...) instead of isinstance?

we use `isinstance(pair, (tuple, list))` usually cause it supports the spack yaml stuff that subtypes builtin types.

kwryankrattiger replied:
the main reason, I didn't know `isinstance` took a tuple of types! TIL, thanks!

haampie commented on lib/spack/spack/mirror.py:
```diff
@@ -271,11 +331,57 @@ def get_url(self, direction: str) -> str:
+    def get_access_pair(self, direction: str) -> Optional[Tuple[str, str]]:
+        pair = self._get_value("access_pair", direction)
+        if type(pair) in (tuple, list):
+            return tuple(map(self._extract_credential_value, pair))
```
This returns a tuple of optional strings, which means your function signature is wrong.

I don't think a function that returns Optional[Tuple[Optional[str], Optional[str]] is great design. Can you improve?

kwryankrattiger replied:
oof, yeah, that is not pretty. The type hint on the return is the intended output, I will clean that up.

haampie replied:
I guess the "can you add typehints" request from above is the reason it's not an error in mypy (?)

kwryankrattiger replied:
all of this is removed, new check is `pair if all(pair) else None`

kwryankrattiger committed:
Address PR comments and test failures

haampie committed:
Fix a few typos and improve error messages
